### PR TITLE
Feat/1481 validate that applicant fulfils residential area rental rules

### PR DESCRIPTION
**Pull request by driatic**
https://dev.azure.com/mimeronline/ONECore/_boards/board/t/Uthyrning/Issues/?workitem=1481

**UPDATE 12 JUNE 2024**

The previous code had a flaw in it were we called onecore-core to get the estateCode for property validation. All property validation is removed from this PR and is in this branch: https://github.com/Bostads-AB-Mimer/onecore-leasing/pull/71. The latter will be rewritten so that it does not call core (creating a circular call chain).

**OLD DESCRIPTION**

This grew much larger than expected. The main reason for this is the `estateCode` field that we need to be able to do the property specific validation. There exists code in property management to get this field and me and David decided to use the code for this to get the estateCode. The result in practice is that leasing now will round trip to onecore-core to get this field from onecore-property-management. 

To query the xpand database for this field was a ton of headache since it requires so many extra joins. It's obvious that we need to spend time migrating static data from xpand (as in rental property data) to onecore. But for now we will do a couple of requests instead inside the oncore ecosystem to get what we need for this validation. 

I did my best to make this as simple as possible but now I'm blind to whatever improvements can be made. Refactoring also seems like a chore at this point but whatever needs to be needs to be done. The tests are quite extensive and manual testing with proper xpand-data will be conducted before this code can be merged. 

Response formats from the new endpoints aren't set in stone. It might be more obvious what we need when the manual tests are performed. 

Sorry for wall of text in this description and in the code 🙏

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -17,4 +17,8 @@ CREDITSAFE__URL=
 
 XPAND_SOAP__USERNAME=
 XPAND_SOAP__PASSWORD=
-XPAND_SOAP__URL=
\ No newline at end of file
+XPAND_SOAP__URL=
+
+CORE__URL=http://localhost:5010
+CORE__USERNAME=
+CORE__PASSWORD=
```
Comment (driatic) [inline .env.template:24]: todo: update onecore-operations with these.

File: src/services/lease-service/index.ts [hunk 2]
```diff
@@ -556,4 +563,95 @@ export const routes = (router: KoaRouter) => {
...
+        ctx.body = {
+          reason:
+            'Applicant already have an active parking space contract in the listings residential area',
+        }
+        ctx.status = 409
+      } catch (error: unknown) {
+        ctx.status = 500
+
+        if (error instanceof Error) {
+          console.log(error.message)
```
Comment (momentiris) [inline src/services/lease-service/index.ts:649]: logger.error({ err: error }, 'error when validating residential rules')

File: src/services/lease-service/tests/index.test.ts [hunk 3]
```diff
@@ -488,4 +490,185 @@ describe('lease-service', () => {
...
+    })
+
+    it('responds with 404 if applicant does not exist', async () => {
+      const listing = ListingFactory.params({
+        districtCode: 'OXB',
+      }).build()
+
+      const getListingSpy = jest
+        .spyOn(listingAdapter, 'getListingById')
+        .mockResolvedValue(listing)
```
Comment (momentiris) [inline src/services/lease-service/tests/index.test.ts:516]: Skulle säga använd .mockResolvedValueOnce överallt när du kör spy returns inuti enskilda tester för att inte riskera att det läcker in i andra tester.

File: src/services/lease-service/tests/rental-rules-validator.test.ts [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        residentialArea: { code: 'XYZ' },
+      }),
+    })
+
+    const result = isHousingContractsOfApplicantInSameAreaAsListing(
+      listing,
+      detailedApplicant
+    )
+
+    expect(result).toBeFalsy()
```
Comment (momentiris) [inline src/services/lease-service/tests/rental-rules-validator.test.ts:64]: Låg risk i det här fallet men med `toBeFalsy()` skulle funktionen i teorin kunna returnera `""` och passera.. :)
```suggestion
    expect(result).toBe(false)
```

File: src/services/lease-service/index.ts [hunk 5]
```diff
@@ -556,4 +563,95 @@ export const routes = (router: KoaRouter) => {
...
+            detailedApplicant
+          )
+
+        if (!doesUserHaveExistingParkingSpaceInSameAreaAsListing) {
+          //applicant is eligible for parking space, applicationType for application should be 'additonal'
+          ctx.body = {
+            reason:
+              'Applicant does not have any active parking space contracts in the listings residential area',
+          }
+          ctx.status = 203 //??
```
[the remark this answers is not in the document]
Comment (driatic) [reply src/services/lease-service/index.ts:634]: 203 borde det ju inte vara men jag tänker att det är gött att särskilja från svaret på rad 584 `special residential area rental rules does not apply to this listing` mot detta svar. Det föregående kanske borde vara 204 och detta 200?

File: src/services/lease-service/index.ts (hunk 2 above)
Comment (driatic) [reply src/services/lease-service/index.ts:649]: Just ja, toppen!

File: src/services/lease-service/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (momentiris) [reply src/services/lease-service/index.ts:634]: Kör på det (om du tror vi behöver särskilja dem två)

File: src/services/lease-service/index.ts (hunk 2 above)
Comment (momentiris) [reply src/services/lease-service/index.ts:649]: Den kanske inte finns i leasing iofs.. 🤷

**Pull request merged**